File: Sources/SRDTestUtilitiesObjC/src/SRDAssertions.m
//
//  SRDAssertions.m
//  SRDTestUtilities
//
//  Copyright © 2021 SomeRandomiOSDev. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "SRDAssertions.h"

// MARK: - Private Function Declarations

static void __SRDAssertAttributedStringContainsAttributeAndValues(NSAttributedString *attributedString, NSAttributedStringKey attributeKey, id attributeValue, id __nullable alternateAttributeValue, NSRange range, const char *filePath, NSUInteger lineNumber, const char *function);

// MARK: - Public Function Definitions

void _SRDAssertAttributedStringContainsAttribute(NSAttributedString *attributedString, NSAttributedStringKey attribute, NSRange range, const char *filePath, NSUInteger lineNumber) {
    if (range.location != NSNotFound && range.length > 0) {
        NSRange attributeRange;
        id value = [attributedString attribute:attribute atIndex:range.location longestEffectiveRange:&attributeRange inRange:NSMakeRange(0, attributedString.length)];

        if (value == nil || !NSEqualRanges(range, attributeRange)) {
            _XCTPreformattedFailureHandler(nil, YES, [NSString stringWithUTF8String:filePath], lineNumber, [NSString stringWithFormat:@"SRDAssertAttributedStringContainsAttribute failed: Expected to find attribute (%@) in range: \"%@\" of attributed string: \"%@\"", attribute, NSStringFromRange(range), attributedString], @"");
        }
    } else {
        __block id value = nil;

        [attributedString enumerateAttributesInRange:NSMakeRange(0, attributedString.length) options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired usingBlock:^(NSDictionary<NSAttributedStringKey, id> *attributes, NSRange _, BOOL *stop) {
            value = attributes[attribute];
            *stop = (value != nil);
        }];

        if (value == nil) {
            _XCTPreformattedFailureHandler(nil, YES, [NSString stringWithUTF8String:filePath], lineNumber, [NSString stringWithFormat:@"SRDAssertAttributedStringContainsAttribute failed: Expected to find attribute (%@) in attributed string: \"%@\"", attribute, attributedString.debugDescription], @"");
        }
    }
}

void _SRDAssertAttributedStringContainsAttributeAndValue(NSAttributedString *attributedString, NSAttributedStringKey attributeKey, id attributeValue, NSRange range, const char *filePath, NSUInteger lineNumber) {
    __SRDAssertAttributedStringContainsAttributeAndValues(attributedString, attributeKey, attributeValue, nil, range, filePath, lineNumber, "SRDAssertAttributedStringContainsAttributeAndValue");
}

void _SRDAssertAttributedStringContainsAttributeAndValues(NSAttributedString *attributedString, NSAttributedStringKey attributeKey, id attributeValue, id __nullable alternateAttributeValue, NSRange range, const char *filePath, NSUInteger lineNumber) {
    __SRDAssertAttributedStringContainsAttributeAndValues(attributedString, attributeKey, attributeValue, alternateAttributeValue, range, filePath, lineNumber, "SRDAssertAttributedStringContainsAttributeAndValues");
}

void _SRDAssertAttributedStringNotContainsAttribute(NSAttributedString *attributedString, NSAttributedStringKey attribute, NSRange range, const char *filePath, NSUInteger lineNumber) {
    __block id value = nil;

    [attributedString enumerateAttributesInRange:(range.location != NSNotFound && range.length > 0) ? range : NSMakeRange(0, attributedString.length) options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired usingBlock:^(NSDictionary<NSAttributedStringKey, id> *attributes, NSRange _, BOOL *stop) {
        value = attributes[attribute];
        *stop = (value != nil);
    }];

    if (value != nil) {
        NSString *description;
        if (range.location != NSNotFound && range.length > 0) {
            description = [NSString stringWithFormat:@"Expected to not find attribute (%@) in range: \"%@\" of attributed string: \"%@\"", attribute, NSStringFromRange(range), attributedString.debugDescription];
        } else {
            description = [NSString stringWithFormat:@"Expected to not find attribute (%@) in attributed string: \"%@\"", attribute, attributedString.debugDescription];
        }

        _XCTPreformattedFailureHandler(nil, YES, [NSString stringWithUTF8String:filePath], lineNumber, [NSString stringWithFormat:@"SRDAssertAttributedStringNotContainsAttribute failed: %@", description], @"");
    }
}

//

void _SRDRegisterUnexpectedFailure(const char *filePath, NSUInteger lineNumber, NSString *message) {
    _XCTPreformattedFailureHandler(nil, NO, [NSString stringWithUTF8String:filePath], lineNumber, message, @"");
}

// MARK: Private Function Definitions

static void __SRDAssertAttributedStringContainsAttributeAndValues(NSAttributedString *attributedString, NSAttributedStringKey attributeKey, id attributeValue, id __nullable alternateAttributeValue, NSRange range, const char *filePath, NSUInteger lineNumber, const char *function) {
    if (range.location != NSNotFound && range.length > 0) {
        NSRange attributeRange;
        id value = [attributedString attribute:attributeKey atIndex:range.location longestEffectiveRange:&attributeRange inRange:NSMakeRange(0, attributedString.length)];

        if (value == nil || !NSEqualRanges(range, attributeRange) || !([value isEqual:attributeValue] || (alternateAttributeValue != nil && [value isEqual:alternateAttributeValue]))) {
            _XCTPreformattedFailureHandler(nil, YES, [NSString stringWithUTF8String:filePath], lineNumber, [NSString stringWithFormat:@"%s failed: Expected to find attribute (%@, \"%@\" or \"%@\") in range: \"%@\" of attributed string: \"%@\"", function, attributeKey, [attributeValue description], [alternateAttributeValue description], NSStringFromRange(range), attributedString], @"");
        }
    } else {
        __block id value = nil;

        [attributedString enumerateAttributesInRange:NSMakeRange(0, attributedString.length) options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired usingBlock:^(NSDictionary<NSAttributedStringKey, id> *attributes, NSRange _, BOOL *stop) {
            value = attributes[attributeKey];
            *stop = (value != nil);
        }];

        if (value == nil || !([value isEqual:attributeValue] || (alternateAttributeValue != nil && [value isEqual:alternateAttributeValue]))) {
            _XCTPreformattedFailureHandler(nil, YES, [NSString stringWithUTF8String:filePath], lineNumber, [NSString stringWithFormat:@"%s failed: Expected to find attribute (%@, \"%@\" or \"%@\") in attributed string: \"%@\"", function, attributeKey, [attributeValue description], [alternateAttributeValue description], attributedString], @"");
        }
    }
}

File: Sources/SRDTestUtilities/SRDAssertions.swift
//
//  SRDAssertions.swift
//  SRDTestUtilities
//
//  Copyright © 2021 SomeRandomiOSDev. All rights reserved.
//

import Foundation
import XCTest

#if SWIFT_PACKAGE
@_exported import SRDTestUtilitiesObjC
#endif // #if SWIFT_PACKAGE

// MARK: Public Functions

/**
 A function for asserting that a given `NSAttributedString` contains a particular
 attribute.

 - Parameters:
   - attributedString: The attributed string over which the assertion is made.
   - attribute: The attribute that the `attributedString` is asserted to contain.
   - range: An optional range in which to specify the _exact_ range in which the
     `attribute` is expected to appear in the `attributedString`. If this parameter
     is `nil`, the `attribute` can appear anywhere in the `attributedString` to
     fulfill the assertion.
 */
public func SRDAssertAttributedStringContainsAttribute(_ attributedString: @autoclosure () -> NSAttributedString, attribute: @autoclosure () -> NSAttributedString.Key, in range: @autoclosure () -> NSRange? = nil, file: StaticString = #file, line: UInt = #line) {
    do {
        try catchingObjCException {
            ___SRDAssertAttributedStringContainsAttribute(attributedString(), attribute(), range() ?? NSRange(location: NSNotFound, length: 0), file.description, line)
        }
    } catch let error as SRDExceptionError {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttribute failed: throwing \\\"\(error.exception.reason ?? "nil")\\\"")
    } catch {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttribute failed: throwing an unknown exception")
    }
}

/**
 A function for asserting that a given `NSAttributedString` contains a particular
 value for a given attribute.

 - Parameters:
   - attributedString: The attributed string over which the assertion is made.
   - attribute: The attribute key and value that the `attributedString` is asserted
     to contain.
   - range: An optional range in which to specify the _exact_ range in which the
     `attributeKey` is expected to appear in the `attributedString`. If this
     parameter is `nil`, the `attributeKey` can appear anywhere in the
     `attributedString` to fulfill the assertion.
 */
public func SRDAssertAttributedStringContainsAttribute<T>(_ attributedString: @autoclosure () -> NSAttributedString, attribute: @autoclosure () -> (key: NSAttributedString.Key, value: T), in range: @autoclosure () -> NSRange? = nil, file: StaticString = #file, line: UInt = #line) where T: Equatable {
    do {
        try catchingObjCException {
            let attribute = attribute()

            ___SRDAssertAttributedStringContainsAttributeAndValue(attributedString(), attribute.key, attribute.value, range() ?? NSRange(location: NSNotFound, length: 0), file.description, line)
        }
    } catch let error as SRDExceptionError {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttributeAndValue failed: throwing \\\"\(error.exception.reason ?? "nil")\\\"")
    } catch {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttributeAndValue failed: throwing an unknown exception")
    }
}

// swiftlint:disable large_tuple
/**
 A function for asserting that a given `NSAttributedString` contains a particular
 value for a given attribute.

 - Parameters:
   - attributedString: The attributed string over which the assertion is made.
   - attribute: The attribute key, value, and optional alternate value that the
     `attributedString` is asserted to contain.
   - range: An optional range in which to specify the _exact_ range in which the
     `attributeKey` is expected to appear in the `attributedString`. If this
     parameter is `nil`, the `attributeKey` can appear anywhere in the
     `attributedString` to fulfill the assertion.

 If the given attribute is found within the `attributedString` (in a given range
 if specified), this function first compares the value for the attribute against
 `attribute.value`. If those are not equal and an alternate value is provided,
 the value is then compared against `attribute.alternateValue`. If both values
 are not equal to the value found in the `attributedString` then a `XCTest`
 assertion is raised.
 */
public func SRDAssertAttributedStringContainsAttribute<T>(_ attributedString: @autoclosure () -> NSAttributedString, attribute: @autoclosure () -> (key: NSAttributedString.Key, value: T, alternateValue: T?), in range: @autoclosure () -> NSRange? = nil, file: StaticString = #file, line: UInt = #line) where T: Equatable {
    do {
        try catchingObjCException {
            let attribute = attribute()

            ___SRDAssertAttributedStringContainsAttributeAndValues(attributedString(), attribute.key, attribute.value, attribute.alternateValue, range() ?? NSRange(location: NSNotFound, length: 0), file.description, line)
        }
    } catch let error as SRDExceptionError {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttributeAndValues failed: throwing \\\"\(error.exception.reason ?? "nil")\\\"")
    } catch {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringContainsAttributeAndValues failed: throwing an unknown exception")
    }
}
// swiftlint:enable large_tuple

/**
 A function for asserting that a given `NSAttributedString` does not contain a
 particular attribute.

 - Parameters:
   - attributedString: The attributed string over which the assertion is made.
   - attribute: The attribute that the `attributedString` is asserted not to
     contain.
   - range: An optional range in which to specify the _exact_ range in which the
     `attribute` is expected not to appear in the `attributedString`. If this
     parameter   is `nil`, the `attribute` must not appear anywhere in the
     `attributedString` to   fulfill the assertion.
 */
public func SRDAssertAttributedStringNotContainsAttribute(_ attributedString: @autoclosure () -> NSAttributedString, attribute: @autoclosure () -> NSAttributedString.Key, in range: @autoclosure () -> NSRange? = nil, file: StaticString = #file, line: UInt = #line) {
    do {
        try catchingObjCException {
            ___SRDAssertAttributedStringNotContainsAttribute(attributedString(), attribute(), range() ?? NSRange(location: NSNotFound, length: 0), file.description, line)
        }
    } catch let error as SRDExceptionError {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringNotContainsAttribute failed: throwing \\\"\(error.exception.reason ?? "nil")\\\"")
    } catch {
        ___SRDRegisterUnexpectedFailure(file.description, line, "SRDAssertAttributedStringNotContainsAttribute failed: throwing an unknown exception")
    }
}
